admin/app/dfp/DfpDataCacheJob.scala

```diff
@@ -67,89 +72,83 @@ object DfpDataCacheJob extends ExecutionContexts with Logging {
-          loadThreshold = Some(threshold),
-          current = cachedLineItems,
-          recentlyAddedIds = Nil,
-          recentlyModifiedIds = Nil,
-          recentlyRemovedIds = Nil
-        )
-
-      def summarizeModifiedCache: LineItemLoadSummary = {
-        def idToLineItem(lineItems: Seq[GuLineItem]) = lineItems.map(item => item.id -> item).toMap
-        val cachedIdToLineItem = idToLineItem(cachedLineItems)
-        val (readyOrDeliveringModified, otherModified) =
-          recentlyModified partition (Seq("READY", "DELIVERING") contains _.status)
-        val readyOrDeliveringIdToLineItem = idToLineItem(readyOrDeliveringModified)
-        val otherModifiedIds = otherModified.map(_.id)
-        val added = readyOrDeliveringIdToLineItem -- cachedIdToLineItem.keys
-        val modified = readyOrDeliveringIdToLineItem filterKeys cachedIdToLineItem.keySet.contains
-        val removed = cachedIdToLineItem filterKeys otherModifiedIds.contains
-        val lineItems = cachedIdToLineItem ++ added ++ modified -- removed.keys
-        LineItemLoadSummary(
-          prevCount = cachedLineItems.size,
-          loadThreshold = Some(threshold),
-          lineItems.values.toSeq.sortBy(_.id),
-          recentlyAddedIds = added.keys,
-          recentlyModifiedIds = modified.keys,
-          recentlyRemovedIds = removed.keys
-        )
-      }
+      // Update existing items with a patch of new items.
+      def updateCachedContent(existingItems: Seq[GuLineItem], newItems: Seq[GuLineItem]): Seq[GuLineItem] = {
 
-      if (recentlyModified.isEmpty) {
-        summarizeUnmodifiedCache
-      } else {
-        summarizeModifiedCache
+        // Create a combined map of all the line items, preferring newer items over old ones (equality is based on id).
+        val updatedLineItemMap = GuLineItem.asMap(existingItems) ++ GuLineItem.asMap(newItems)
+
+        // These are the existing, cached keys.
+        val existingKeys = existingItems.map(_.id).toSet
+
+        val (active, inactive) = newItems partition (Seq("READY", "DELIVERING", "DELIVERY_EXTENDED") contains _.status)
+        val activeKeys = active.map(_.id).toSet
+        val inactiveKeys = inactive.map(_.id).toSet
+
+        val added = activeKeys -- existingKeys
+        val modified = activeKeys intersect existingKeys
+        val removed = inactiveKeys intersect existingKeys
+
+        // New cache contents.
+        val updatedKeys = existingKeys ++ added -- removed
+
+        updatedKeys.toSeq.sorted.map(updatedLineItemMap)
       }
-    }
 
-    if (cachedLineItems.isEmpty) {
-      summarizeNewCache
-    } else {
-      summarizeUpdatedCache
+      LineItemLoadSummary(
+        prevCount = cachedLineItems.validItems.size,
+        loadThreshold = Some(threshold),
+        validLineItems = updateCachedContent(cachedLineItems.validItems, recentlyModified.validItems),
+        invalidLineItems = updateCachedContent(cachedLineItems.problemItems, recentlyModified.problemItems),
```

Comment:
Should we rename `cachedLineItem.problemItems` to `cachedLineItem.invalidItems`?